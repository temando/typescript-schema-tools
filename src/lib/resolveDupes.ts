export interface ISchemaShape {
  $id?: string;
  id?: string;
  definitions?: { [k: string]: ISchemaShape };
  $schema?: string;
  type?: string;
  $ref?: string;
  properties?: { [k: string]: ISchemaShape };
  [k: string]: any;
}

const isRootSchema = (schema: ISchemaShape) => Boolean(schema.id || schema.$id);

const simplifySchema = (schema: ISchemaShape) => {
  const cleaned: ISchemaShape = { ...schema };
  delete cleaned.$schema;
  delete cleaned.id;
  delete cleaned.$id;
  delete cleaned.definitions;

  return cleaned;
};

const isSchemaADupe = (target: ISchemaShape, subject: ISchemaShape): boolean => {
  if (target === subject) { return false; }

  if (isRootSchema(target) || !isRootSchema(subject)) { return false; }

  const simpleSubject = simplifySchema(subject);

  const targetStr = JSON.stringify(target);
  const subjectStr = JSON.stringify(simpleSubject);

  const anyIsEmptyObject = targetStr === '{}';

  return !anyIsEmptyObject && targetStr === subjectStr;
};

const findDupeSchema = (target: ISchemaShape, schemas: ISchemaShape[]): ISchemaShape|undefined => {
  for (const schema of schemas) {
    if (isSchemaADupe(target, schema)) {
      return schema;
    }
  }
};

export function resolveDupes (schemas: ISchemaShape[]) {
  let rootSchemas = schemas.filter(isRootSchema);

  const recursivelyReplaceDupes = (obj: ISchemaShape): ISchemaShape => {
    if (obj && obj instanceof Object) {
      const schemaRoot = findDupeSchema(obj, schemas);

      if (schemaRoot) {
        return { $ref: schemaRoot.$id || schemaRoot.id };
      } else {
        obj = obj instanceof Array ? [...obj] : { ...obj };

        Object.keys(obj).forEach((key) => {
          obj[key] = recursivelyReplaceDupes(obj[key]);
        });
      }
    }

    return obj;
  };

  rootSchemas = rootSchemas.map(recursivelyReplaceDupes);
  const newSchemas = schemas.map(recursivelyReplaceDupes);

  return newSchemas;
}
