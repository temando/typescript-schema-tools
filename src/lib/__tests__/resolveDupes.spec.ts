import { resolveDupes } from '../resolveDupes';

const dupedSchema = {
  type: 'object',
  properties: {
    a: {
      type: 'string',
      enum: ['a', 'b', 'c'],
    },
    b: { type: 'string' },
    c: {
      type: 'string',
      enum: ['a', 'b', 'c'],
    },
  },
  required: ['a'],
};

const rootSchemas = [
  {
    type: 'string',
    enum: ['a', 'b', 'c'],
    definitions: {},
    $schema: 'http://json-schema.org/draft-07/schema#',
    $id: 'wew',
  },
];

it('resolves dupes in a schema to a root', () => {
  const deDuped = resolveDupes([dupedSchema, ...rootSchemas]);

  expect(deDuped).toMatchSnapshot();
});
