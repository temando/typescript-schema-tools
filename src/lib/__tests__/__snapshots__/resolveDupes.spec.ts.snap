// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolves dupes in a schema to a root 1`] = `
Array [
  Object {
    "properties": Object {
      "a": Object {
        "$ref": "wew",
      },
      "b": Object {
        "type": "string",
      },
      "c": Object {
        "$ref": "wew",
      },
    },
    "required": Array [
      "a",
    ],
    "type": "object",
  },
  Object {
    "$id": "wew",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": Object {},
    "enum": Array [
      "a",
      "b",
      "c",
    ],
    "type": "string",
  },
]
`;
